src/Aspire.Hosting/ApplicationModel/CustomResourceLoggerAnnotation.cs

```diff
@@ -0,0 +1,102 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Threading.Channels;
+using Aspire.Dashboard.Otlp.Storage;
+using Microsoft.Extensions.Logging;
+
+namespace Aspire.Hosting.ApplicationModel;
+
+/// <summary>
+/// A annotation that exposes a Logger for the resource to write to.
+/// </summary>
+public sealed class CustomResourceLoggerAnnotation : IResourceAnnotation
+{
+    private readonly ResourceLogger _logger;
+
+    // History of logs, capped at 10000 entries.
+    private readonly CircularBuffer<(string Content, bool IsErrorMessage)> _backlog = new(10000);
+
+    /// <summary>
+    /// Creates a new <see cref="CustomResourceLoggerAnnotation"/>.
+    /// </summary>
+    public CustomResourceLoggerAnnotation()
+    {
+        _logger = new ResourceLogger(this);
+    }
+
+    /// <summary>
+    /// Watch for changes to the log stream for a resource.
+    /// </summary>
+    /// <returns> The log stream for the resource. </returns>
+    public IAsyncEnumerable<IReadOnlyList<(string Content, bool IsErrorMessage)>> WatchAsync() => new LogAsyncEnumerable(this);
+
+    // This provides the fan out to multiple subscribers.
+    private Action<(string, bool)>? OnNewLog { get; set; }
+
+    /// <summary>
+    /// The logger for the resource to write to. This will write updates to the live log stream for this resource.
+    /// </summary>
+    public ILogger Logger => _logger;
```

Comment:
This allows writing logs but there isn't a way to say logs are finished. Should there be?

Reply:
That's interesting. Not sure it needs to be a different API as users can just log now?

Reply:
To be honest I don't know if it's needed. Does the watch console logs stream end when an app stops?

Reply:
Hmmm, thinking about it more, this is a little bit tricky because it's one to many fan out and each channel ends when the subscriber closes.

Reply:
OK fixing.